### PR TITLE
DRY Refactoring

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR extracts similar segments from the code.
Also: Lot's of refactoring, namely we use Streams now to parse point files, and regex instead of some weird token logic + manual algorithms.

This PR is WIP, because I haven't added any tests yet, but feel encouraged to review it now already.
Tests will follow tomorrow but it's getting quite late here 😅 

BTW: Haven't really tested this, but I'm not noticing any performance differences regarding the regex change. I'd appreciate of someone could verify this on their machine as well.

### CODE REVIEW

**Comment by ron-murhammer** [comment]
Looking pretty good so far and I'm sure you can see now why I decided it was better to look to refactor this mess separately from the feature PR.  I'd make sure you test a handful of maps as well as the unit tests just to make sure nothing funky happens.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 1]
```diff
@@ -31,6 +31,11 @@
  * String -> a list of points, or string-> list of polygons.
  */
 public final class PointFileReaderWriter {
+
+  private static final Pattern pointPattern = Pattern.compile("\\s*\\(\\s*(\\d+)\\s*,\\s*(\\d+)\\s*\\)");
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: If we have the knowledge, some examples of what these would match would be excellent to document.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 2]
```diff
@@ -256,115 +199,62 @@ public static void writeOneToManyPlacements(final OutputStream sink,
...
+    while (matcher.find()) {
+      points.add(new Point(Integer.parseInt(matcher.group(1)), Integer.parseInt(matcher.group(2))));
     }
     mapping.put(name, points);
     return Tuple.of(name, points);
   }
+
+  private static void readStream(final InputStream stream, Consumer<String> lineParser)
+      throws IOException {
+    try (Reader inputStreamReader = new InputStreamReader(new CloseShieldInputStream(stream), StandardCharsets.UTF_8);
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: Does `new CloseShieldInputStream(stream)` need to be closed?

**Comment by DanVanAtta** [comment]
@ron-murhammer I understand - my concern is that previously we did not do much to address the code tangle problem, only added to it. A simple TODO comment to point out the locations to keep consistent and potentially refactor would have been a good minimum and a step to reducing the code tangle problem.

**Comment by DanVanAtta** [comment]
@RoiEXLab i'd recommend to do the DRY as one refactor; the update of logic would be best done with support of proper testing and IMO a second pass and PR.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java (hunk 2 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: `new CloseShieldInputStream(stream)` is a wrapper that prevents the underlying stream to be closed, even if close is being called.
So no.

**Comment by RoiEXLab** [comment]
@DanVanAtta 
> @RoiEXLab i'd recommend to do the DRY as one refactor; the update of logic (" and regex instead of some weird token logic + manual algorithms.") would be best done with support of proper testing and IMO a second pass and PR. 

I disagree. There are already lots of Tests in place thanks to @ssoloff in #2889 and therefore everything is well-tested already. (An exception to this is the newly introduced feature by @ron-murhammer) so this PR is unlikely to break something + this is a heavily used utility class that is needed to load a game correctly, so if we just start a lot of maps and don't get any errors we are probably fine because the syntax is really simple.

**Comment by RoiEXLab** [comment]
I added some tests and got rid of some "potential issues" that made a small difference compared to the previous implementation (like permitting empty names etc.).
Besides that I noticed some unintentional behaviour in @ron-murhammer's overflow implementation.
The biggest issue was an inconsistency in a couple of cases:
1. The "attribute name" is never checked, typos are likely. Problematic if we ever want to add another "attribute"
2. ` | overflowToLeft=true ` actually sets it to false because `true ` != `true` (extra space at the end)
3. While ` | overflowToLeft=a b c d e f` is perfectly fine ` | overflowToLeft=` results in an ArrayIndexOutOfBoundsException

@ron-murhammer Please look at my test cases for this new method and try to improve the method in a follow up PR. I don't know what the maps makers have been doing with this feature already.

PR is now ready for review again. @DanVanAtta

**Comment by ron-murhammer** [comment]
@RoiEXLab Not really following your comments.

**Comment by DanVanAtta** [comment]
@RoiEXLab I recommend splitting the logic updates to a new PR The straight forward stream updates look good, but the deeper cleanups need more time to inspect.

The upates ins  `readMultiplePolygons` are looking good,  only about 5% reviewed, will come back to this hopefully tomorrow or day after.

**Comment by RoiEXLab** [comment]
@ron-murhammer Sorry if I wasn't clear enough:
I was talking about the logic in the oneToManyPlacements method you created in order to implement the overflow to the left functionality.
The examples I provided were about how this additional functionality behaved unexpectedly like exceptions when similar constructs worked fine and partial whitespace insensitivity.
So if you could have a look again at this code I'd appreciate this.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 3]
```diff
@@ -40,30 +50,21 @@ private PointFileReaderWriter() {}
...
-    final StringTokenizer tokens = new StringTokenizer(line, "", false);
-    final String name = tokens.nextToken("(").trim();
-    if (mapping.containsKey(name)) {
-      throw new IOException("name found twice:" + name);
+  private static void readSingle(final String line, final Map<String, Point> mapping) {
+    final Matcher matcher = singlePointPattern.matcher(line);
+    if (matcher.find()) {
+      final String name = matcher.group(1);
+      if (mapping.containsKey(name)) {
+        throw new IllegalArgumentException("name found twice:" + name);
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: nit: missing a space after `:`

Should we also log the `mapping` parameter here too?

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 4]
```diff
@@ -40,30 +50,21 @@ private PointFileReaderWriter() {}
...
+      }
+      mapping.put(name, new Point(Integer.parseInt(matcher.group(2)), Integer.parseInt(matcher.group(3))));
+    } else {
+      throw new IllegalArgumentException("Invalid Point Pattern");
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: Can we add some info here so we know which point pattern was invalid?

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 5]
```diff
@@ -40,30 +50,21 @@ private PointFileReaderWriter() {}
...
 
-  private static void readSingle(final String line, final Map<String, Point> mapping) throws IOException {
-    final StringTokenizer tokens = new StringTokenizer(line, "", false);
-    final String name = tokens.nextToken("(").trim();
-    if (mapping.containsKey(name)) {
-      throw new IOException("name found twice:" + name);
+  private static void readSingle(final String line, final Map<String, Point> mapping) {
+    final Matcher matcher = singlePointPattern.matcher(line);
+    if (matcher.find()) {
+      final String name = matcher.group(1);
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: While we have context, if you can, are there any better variable names for this `name` variable? Name of what?

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java [hunk 6]
```diff
@@ -105,32 +97,36 @@ public static void writeOneToManyPolygons(final OutputStream sink, final Map<Str
...
 
-  private static void write(final StringBuilder buf, final OutputStream sink) throws IOException {
+  private static void write(final String string, final OutputStream sink) throws IOException {
     try (Writer out = new OutputStreamWriter(new CloseShieldOutputStream(sink), StandardCharsets.UTF_8)) {
-      out.write(buf.toString());
+      out.write(string);
     }
   }
 
+  private static String pointsToString(final List<Point> points) {
```
Comment (DanVanAtta) [inline game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: nit, first usage is around line 140, private methods should follow after their first usage, please move to be below the calling method.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java (hunk 3 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: Not the complete mapping, but the existent value for the key to be able to Ctrl+F it.

File: game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java (hunk 4 above)
Comment (RoiEXLab) [reply game-core/src/main/java/games/strategy/util/PointFileReaderWriter.java]: Good catch. I wanted to change that but I forgot 😅

**Comment by RoiEXLab** [comment]
@DanVanAtta Updated, tell me what you think

File: game-core/src/test/java/games/strategy/util/PointFileReaderWriterTest.java [hunk 7]
```diff
@@ -107,6 +160,103 @@ public void shouldReadMultiplePointsPerName() throws Exception {
...
+    }
+
+    @Test
+    public void shouldReturnEmptyMapWhenStreamIsEmpty() throws Exception {
+      assertThat(IoUtils.readFromMemory(new byte[0], PointFileReaderWriter::readOneToManyPlacements),
+          is(Collections.emptyMap()));
+    }
+
+    @Test
+    public void shouldReadMultiplePlacementsPerName() throws Exception {
```
Comment (ron-murhammer) [inline game-core/src/test/java/games/strategy/util/PointFileReaderWriterTest.java]: @RoiEXLab The tests here look good to me. Essentially the only time "overflowToLeft" should be true is if the line ends with " | overflowToLeft=true". Any other cases, it should be false. We also control the writing of that text in the map creator so as long as the 2 match up, we shouldn't have any issues.

**Comment by RoiEXLab** [comment]
@DanVanAtta Merging this as I adressed all of your suggestions.

**Pull request merged**